File: src/main/java/redis/RedisProtocol.java
package redis;

import redis.reply.BulkReply;
import redis.reply.ErrorReply;
import redis.reply.IntegerReply;
import redis.reply.MultiBulkReply;
import redis.reply.Reply;
import redis.reply.StatusReply;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;

/**
 * Implements the Redis Universal Protocol. Send a command, receive a command, send a reply
 * and receive a reply.
 */
public class RedisProtocol {

  private DataInputStream is;
  private OutputStream os;

  public RedisProtocol(Socket socket) throws IOException {
    is = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
    os = new BufferedOutputStream(socket.getOutputStream());
  }

  public static byte[] readBytes(DataInputStream is) throws IOException {
    int size = Integer.parseInt(is.readLine());
    if (size == -1) {
      return null;
    }
    byte[] bytes = new byte[size];
    int total = 0;
    int read;
    while (total < bytes.length && (read = is.read(bytes, total, bytes.length - total)) != -1) {
      total += read;
    }
    if (total < bytes.length) {
      throw new IOException("Failed to read enough bytes: " + total);
    }
    int cr = is.read();
    int lf = is.read();
    if (cr != '\r' || lf != '\n') {
      throw new IOException("Improper line ending: " + cr + ", " + lf);
    }
    return bytes;
  }

  public static Reply receive(DataInputStream is) throws IOException {
    int code = is.read();
    switch (code) {
      case StatusReply.MARKER: {
        return new StatusReply(is.readLine());
      }
      case ErrorReply.MARKER: {
        return new ErrorReply(is.readLine());
      }
      case IntegerReply.MARKER: {
        return new IntegerReply(Integer.parseInt(is.readLine()));
      }
      case BulkReply.MARKER: {
        byte[] bytes = readBytes(is);
        return new BulkReply(bytes);
      }
      case MultiBulkReply.MARKER: {
        int size = Integer.parseInt(is.readLine());
        byte[][] byteArrays = new byte[size][];
        for (int i = 0; i < size; i++) {
          if (is.read() == BulkReply.MARKER) {
            byteArrays[i] = readBytes(is);
          } else {
            throw new IOException("Unexpected character in stream");
          }
        }
        return new MultiBulkReply(byteArrays);
      }
      default: {
        throw new IOException("Unexpected character in stream: " + code);
      }
    }
  }

  public Reply send(Command command) throws IOException {
    command.write(os);
    os.flush();
    return receive(is);
  }

  public Command receive() throws IOException {
    return Command.read(is);
  }

  public void send(Reply reply) throws IOException {
    reply.write(os);
    os.flush();
  }

  public void write(byte[] bytes) throws IOException {
    os.write(bytes);
    os.flush();
  }
}

File: go/src/github.com/spullara/redis/redisprotocol.go
/**
 * TODO: Edit this
 *
 * User: sam
 * Date: 5/19/13
 * Time: 5:44 PM
 */
package redis

import "io"
import (
	"errors"
	"fmt"
	"bufio"
)

const (
	StatusMarker    = '+'
	BulkMarker      = '$'
	IntegerMarker   = ':'
	MultiBulkMarker = '*'
	ErrorMarker     = '-'
)

type Reply interface {
}

type SimpleReply struct {
	bytes []byte
}

type StatusReply SimpleReply
type ErrorReply SimpleReply
type BulkReply SimpleReply

type IntegerReply struct {
	integer int64
}

type MultiBulkReply struct {
	replies []Reply
}

func (reply *StatusReply) Status() (status string) {
	return string(reply.bytes)
}

func (reply *ErrorReply) Error() (err string) {
	return string(reply.bytes)
}

func Receive(in io.Reader) (reply Reply, err error) {
	br := bufio.NewReader(in)
	code, err := br.ReadByte()
	if err != nil {
		return nil, err
	}
	switch code {
	case StatusMarker:
		line, _, err := br.ReadLine()
		if err != nil {
			return nil, err
		}
		return &StatusReply {
			bytes: line,
		}, nil
	case ErrorMarker:
		line, _, err := br.ReadLine()
		if err != nil {
			return nil, err
		}
		return &ErrorReply {
			bytes: line,
		}, nil
	case IntegerMarker:
		integer, err := readLong(br)
		if err != nil {
			return nil, err
		}
		return &IntegerReply {
			integer: integer,
		}, nil
	case BulkMarker:
		bytes, err := readBytes(br)
		if err != nil {
			return nil, err
		}
		return &BulkReply {
			bytes: bytes,
		}, nil
	case MultiBulkMarker:
		size, err := readLong(br)
		if err != nil {
			return nil, err
		}
		if size == -1 {
			return &MultiBulkReply {
				replies: nil,
			}, nil
		} else {
			if size < 0 {
				return nil, errors.New("Invalid negative size")
			}
			replies := make([]Reply, size)
			for i := 0; i < int(size); i++ {
				replies[i], err = Receive(br)
				if err != nil {
					return nil, err
				}
			}
			return &MultiBulkReply {
				replies: replies,
			}, nil
		}
		return nil, nil
	default:
		return nil, errors.New("Unexpected character in stream")
	}
	panic("Should not reach here")
}


func readBytes(in io.Reader) (bytes []byte, err error) {
	br := bufio.NewReader(in)
	size, err := readLong(br)
	if err != nil {
		return nil, err
	}
	if size == -1 {
		return nil, nil
	}
	if size < 0 {
		return nil, errors.New("Invalid size: " + fmt.Sprint("%d", size))
	}
	bytes = make([]byte, int(size))
	lin := io.LimitReader(br, size)
	for total := 0; total < int(size); {
		read, err := lin.Read(bytes[total:size])
		if err != nil {
			return nil, err
		}
		total += read
	}
	cr, err := br.ReadByte()
	if err != nil {
		return nil, err
	}
	lf, err := br.ReadByte()
	if err != nil {
		return nil, err
	}
	if cr != 13 || lf != 10 {
		return nil, errors.New("Improper line ending")
	}
	return
}

func readLong(in *bufio.Reader) (result int64, err error) {
	read, err := in.ReadByte()
	if err != nil {
		return -1, err
	}
	var sign int
	if read == '-' {
		read, err = in.ReadByte()
		if err != nil {
			return -1, err
		}
		sign = -1
	} else {
		sign = 1
	}
	var number int64
	for number = 0; err == nil; read, err = in.ReadByte() {
		if read == 13 {
			read, err = in.ReadByte()
			if err != nil {
				return -1, err
			}
			if read == 10 {
				return number*int64(sign), nil
			} else {
				return -1, errors.New("Bad line ending")
			}
		}
		value := read - '0'
		if value >= 0 && value < 10 {
			number *= 10
			number += int64(value)
		} else {
			return -1, errors.New("Invalid digit")
		}
	}
	return -1, err
}

